Python: Implement Support for DynamoDB Catalog

arminnajafi opened the pull request:
Fixes #6541

arminnajafi commented:
@Fokko @jackye1995 
Ready for review

jackye1995 commented on python/mkdocs/docs/configuration.md:
````diff
@@ -85,3 +85,15 @@ catalog:
+```yaml
+catalog:
+  default:
+    type: dynamodb
````
I think table name is required for dynamoDB catalog

arminnajafi replied:
Added.

jackye1995 commented on python/pyiceberg/catalog/base_aws_catalog.py:
```diff
@@ -0,0 +1,163 @@
+EXTERNAL_TABLE_TYPE = "EXTERNAL_TABLE"
+
+
+class BaseAwsCatalog(Catalog, ABC):
```
I am not sure if we need to extract this base class. I agree there are common logic shared across catalogs, but I don't think there is too much in common between Dynamo and Glue catalog implementations. 

I remember we have talked about having a base.py class when doing glue catalog implementation, but I think we ended up not doing that and just added util functions for shared functionalities. Maybe we should continue with that pattern.

arminnajafi replied:
Removed `BaseAwsCatalog` class. I moved things around a bit to still remove duplicate code. There is some duplicate code left in `create_table()` but I think it's fine.

jackye1995 commented on python/pyiceberg/catalog/hive.py:
```diff
@@ -548,10 +511,9 @@ def update_namespace_properties(
-        return PropertiesUpdateSummary(
-            removed=list(removed or []), updated=list(updates.keys() if updates else []), missing=list(expected_to_change)
-        )
+        return PropertiesUpdateSummary(removed=list(removed or []), updated=list(updated or []), missing=list(expected_to_change))
```
I see there are multiple places with this kind of style change and newline additions. If not required by linter, can we avoid these changes?

arminnajafi replied:
This is done my `make lint`. I changed it to the previous formatting. `make lint` put it back in one line. 
I didn't find another example like this. Please point them out and I'll fix it.

JonasJ-ap commented on python/pyiceberg/catalog/__init__.py:
```diff
@@ -431,3 +440,114 @@ def namespace_from(identifier: Union[str, Identifier]) -> Identifier:
+    def _get_metadata_location(location: str) -> str:
+        return f"{location}/metadata/00000-{uuid.uuid4()}.metadata.json"
+
+    def _get_updated_props_and_update_summary(
```
[Doubt] May be this method can also be utilized by `update_namespace_properties` in `hive.py`?

arminnajafi replied:
Thanks for seeing this. Yeah that was my plan too. 

But hive is doing this slightly different:

`hive.py`:
```
if removals:
    for key in removals:
        if key in parameters:
            parameters[key] = None
            removed.add(key)
```

`glue.py` and `dynamodb.py`:
```
if removals:
    for key in removals:
        if key in updated_properties:
            updated_properties.pop(key)
            removed.add(key)
```

The difference is in `parameters[key] = None` vs `updated_properties.pop(key)` I wasn't sure if this difference was intended or we could consolidate the behavior and refactor hive too.

Hopping @Fokko can clarify that.

arminnajafi replied:
So I keep it as is fro now and resolve the thread.

JonasJ-ap reviewed (commented):
@arminnajafi  Thank you for the implementation of DynamoDB Catalog and the refactor to other catalogs.

arminnajafi commented:
@JonasJ-ap 
Thanks for reviewing this PR and also your Glue Catalog PR. This work is mostly based on that PR.

arminnajafi commented on python/tests/catalog/__init__.py:
```diff
@@ -0,0 +1,52 @@
+#  under the License.
+
+
+import random
```
Ack. Thanks for letting me know. This `__init__.py` is removed.

arminnajafi commented on python/pyiceberg/catalog/dynamodb.py:
```diff
@@ -0,0 +1,733 @@
+            self.dynamodb.exceptions.ItemCollectionSizeLimitExceededException,
+            self.dynamodb.exceptions.TransactionConflictException,
+            self.dynamodb.exceptions.RequestLimitExceeded,
+            self.dynamodb.exceptions.InternalServerError,
```
Yeah this is something that we can discuss as a team and decide what's the best behavior we want to provide. 
Generally I think it's a good idea for an app to catch all its raw exceptions and re-throw with its own format. Sometimes it's needed to change the format or add more details or sometimes even to hide the underlying raw exception from the end user. 

But in this case I just wanted to be explicit about all of the exception that each of these functions throw, re-throw them as our own exception for now, and and decide later as a team what we want to do with them.

arminnajafi commented on python/pyiceberg/catalog/dynamodb.py:
```diff
@@ -0,0 +1,733 @@
+        """
+
+        if namespace:
+            raise ValidationError("This API is not supported for hierarchical namespaces.")
```
This is a good point. 

I actually didn't think just returning an empty list would be a good idea. Because in that case if a user sends in a namespace they will just get an empty list. That will imply their dataset does not have any database under their provided namespace which which may not be the truth and without getting any exception from the app saying that this feature is not supported and it's not like the answer to your question is an empty list. 

I like to hear @jackye1995 and @Fokko option on this too and decide whether we want to go with an empty list or throw.

arminnajafi replied:
Discussion here:
https://github.com/apache/iceberg/pull/6646#discussion_r1097947897

Fokko commented on python/pyiceberg/catalog/dynamodb.py:
```diff
@@ -0,0 +1,776 @@
+    regular_json = {}
+    for column_name, val_dict in dynamo_json.items():
+        keys = list(val_dict.keys())
+        assert len(keys) == 1
```
We tend to use assertions only in test code since they can be disabled. Let's throw a ValueError instead.

arminnajafi replied:
Done.

Fokko reviewed (commented):
Hey @arminnajafi! First of all, sorry for my super late reply to this. I was traveling last week, and this PR needs some concentration. I also had to set up a dynamodb catalog, and got it working:

```
➜  arrow git:(fd-fix-drop-docstring) ✗ python3
Python 3.10.9 (main, Dec 15 2022, 17:11:09) [Clang 14.0.0 (clang-1400.0.29.202)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> from pyiceberg.catalog import load_catalog
>>> from pyiceberg.expressions import GreaterThanOrEqual, And, LessThan
>>> 
>>> catalog = load_catalog('dynamo')
>>> 
>>> tbl = catalog.load_table('nyc.taxis')
>>> 
>>> tbl.scan(row_filter=And(
...     GreaterThanOrEqual("tpep_pickup_datetime", "2021-04-01T00:00:00.000000+00:00"),
...     LessThan("tpep_pickup_datetime", "2021-05-01T00:00:00.000000+00:00")
... )).to_arrow().to_pandas()
Empty DataFrame
Columns: [VendorID, tpep_pickup_datetime, tpep_dropoff_datetime, passenger_count, trip_distance, RatecodeID, store_and_fwd_flag, PULocationID, DOLocationID, payment_type, fare_amount, extra, mta_tax, tip_amount, tolls_amount, improvement_surcharge, total_amount, congestion_surcharge]
Index: []
>>> tbl = catalog.load_table('nyc.taxis')
>>> tbl.scan().to_arrow().to_pandas()
       VendorID tpep_pickup_datetime tpep_dropoff_datetime passenger_count  ... tolls_amount improvement_surcharge total_amount congestion_surcharge
0             1  2020-04-01 00:41:22   2020-04-01 01:01:53               1  ...            0                   0.3          6.8                    0
1             1  2020-04-01 00:56:00   2020-04-01 01:09:25               1  ...            0                   0.3        16.55                    0
2             1  2020-04-01 00:00:26   2020-04-01 00:09:25               1  ...            0                   0.3         14.8                  2.5
3             1  2020-04-01 00:24:38   2020-04-01 00:34:38               0  ...            0                   0.3         14.8                  2.5
4             2  2020-04-01 00:13:24   2020-04-01 00:18:26               1  ...            0                   0.3         13.3                  2.5
...         ...                  ...                   ...             ...  ...          ...                   ...          ...                  ...
237988     None  2020-04-30 19:47:00   2020-04-30 20:07:00            None  ...            0                   0.3        29.62                    0
237989     None  2020-04-30 19:34:00   2020-04-30 19:39:00            None  ...            0                   0.3         5.85                    0
237990     None  2020-04-30 19:58:14   2020-04-30 20:03:47            None  ...            0                   0.3        10.89                  2.5
237991     None  2020-04-30 19:35:15   2020-04-30 19:56:51            None  ...         6.12                   0.3        31.84                  2.5
237992     None  2020-04-30 19:27:47   2020-04-30 19:45:00            None  ...            0                   0.3        17.21                    0

[237993 rows x 18 columns]
```

I've done some basic tests and will dig into the code in detail tomorrow. But it is looking great already

Fokko commented on python/pyiceberg/catalog/dynamodb.py:
```diff
@@ -0,0 +1,776 @@
+
+    def _ensure_catalog_table_exists_or_create(self) -> None:
+        if self._dynamodb_table_exists():
+            return
```
```suggestion
            return None
```

arminnajafi replied:
I think empty return is the same as return None. But I added it anyway to be more explicit.

Fokko commented on python/pyiceberg/catalog/dynamodb.py:
```diff
@@ -0,0 +1,776 @@
+        try:
+            self.drop_table(from_identifier)
+        except (NoSuchTableError, GenericDynamoDbError) as e:
+            self.drop_table(to_identifier)
```
This can throw an exception as well, should we swallow that one?

arminnajafi replied:
Added a new try catch

Fokko commented on python/pyiceberg/catalog/dynamodb.py:
```diff
@@ -0,0 +1,776 @@
+        except self.dynamodb.exceptions.ConditionalCheckFailedException as e:
+            raise NoSuchNamespaceError(f"Database does not exist: {database_name}") from e
+        except (
+            self.dynamodb.exceptions.ProvisionedThroughputExceededException,
```
Should we generalize these blocks? Since we're catching the same exceptions all the time. WDYT?

arminnajafi replied:
Done.

arminnajafi replied:
I tried to doa function, also a variable:
```
ddb_put_and_delete_exceptions: Tuple[Exception, ...] = (
            self.dynamodb.exceptions.ResourceNotFoundException,
            self.dynamodb.exceptions.ItemCollectionSizeLimitExceededException,
            self.dynamodb.exceptions.TransactionConflictException,
        )
```

But mypy fails me:
```
mypy.....................................................................Failed
- hook id: mypy
- exit code: 1

python/pyiceberg/catalog/dynamodb.py:358: error: Exception type must be derived from BaseException  [misc]
...
```

I think it's not compatible somehow. I reverting this change. But please let me know if there is a good way to do this.

Fokko commented on python/pyiceberg/catalog/dynamodb.py:
```diff
@@ -0,0 +1,776 @@
+        file = io.new_input(metadata_location)
+        metadata = FromInputFile.table_metadata(file)
+        return Table(
+            identifier=(database_name, table_name),
```
This should also include the name of the catalog:
```suggestion
            identifier=(self.name, database_name, table_name),
```

arminnajafi replied:
I added this for dynamo but it doesn't seem to be consistent with Glue and Hive. 
Please let me know if those are supposed to be different.

Fokko replied:
No, this should be the case for all of the catalogs

arminnajafi replied:
Done. 

To my knowledge hive and glue are not following this pattern. Please let me know if you want to apple the same change to glue and hive as well.

jackye1995 replied:
If that's the case we can raise a follow up PR to address this inconsistency

arminnajafi replied:
Cool.

Fokko commented on python/pyiceberg/catalog/dynamodb.py:
```diff
@@ -0,0 +1,776 @@
+def _get_update_database_item(namespace_item: Dict[str, Any], updated_properties: Properties) -> Dict[str, Any]:
+    current_timestamp_ms = str(round(time() * 1000))
+
+    # noinspection PyDictCreation
```
Just curious, why not initialize the dict right away instead of adding the items one by one?

arminnajafi replied:
Done. I though it's more readable that way but if I had to add `# noinspection PyDictCreation` for it, it's probably not a good sign.

Fokko commented on python/pyiceberg/catalog/dynamodb.py:
```diff
@@ -0,0 +1,776 @@
+    return _dict
+
+
+def _get_create_catalog_attribute_definitions() -> List[Dict[str, str]]:
```
instead of methods, we could make these also constants.

arminnajafi replied:
Done.

Fokko commented on python/pyiceberg/catalog/dynamodb.py:
```diff
@@ -0,0 +1,776 @@
+
+
+def _get_namespace_properties(namespace_dict: Dict[str, str]) -> Properties:
+    namespace_properties = {}
```
I'm a bit of a comprehension geek, but what do you think of:
```python
return [
    _remove_property_prefix(key): val
    for key, val in namespace_dict.items()
    if key.startswith(PROPERTY_KEY_PREFIX)
]
```

arminnajafi replied:
Nice. Yeah looks much better. 
:)

Changed list comprehension to dict comprehension though in your code snippet.

Fokko commented on python/pyiceberg/exceptions.py:
```diff
@@ -90,3 +90,19 @@ class SignError(Exception):
+    pass
+
+
+class ItemNotFound(DynamoDbError):
```
What do you think of replacing this one with the build in ValueError? It is not being used a lot, and we want to be mindful of introducing new exceptions.

arminnajafi replied:
Done.

Fokko reviewed (commented):
Just a few more comments, but it is looking great! Thanks @arminnajafi for working on this, and special thanks for consolidating overlapping logic between the different catalogs 👏🏻

arminnajafi commented on python/mkdocs/docs/configuration.md:
```diff
@@ -85,3 +85,16 @@ catalog:
+catalog:
+  default:
+    type: dynamodb
+    dynamodb_table_name: iceberg
```
Of course. Good point. Sorry for the miss.

arminnajafi commented on python/pyiceberg/catalog/dynamodb.py:
```diff
@@ -0,0 +1,776 @@
+        except ConditionalCheckFailedException as e:
+            raise TableAlreadyExistsError(f"Table {database_name}.{table_name} already exists") from e
+
+        loaded_table = self.load_table(identifier=identifier)
```
Done.

arminnajafi commented on python/pyproject.toml:
```diff
@@ -103,6 +103,7 @@ hive = ["thrift"]
 s3fs = ["s3fs"]
 glue = ["boto3"]
 adlfs = ["adlfs"]
+dynamodb = ["boto3"]
```
Can you please explain this a little more? I don't think I completely get this hierarchical namespace thing. 

Before that, can please also share the repro steps? How did you load nyc database, your local config, and ...

arminnajafi replied:
In the meanwhile I made it the same as Glue and return an empty list.

JonasJ-ap replied:
I think it is related to CLI's `list` logic:
https://github.com/apache/iceberg/blob/775b3ce3a7421a9265b3fa7ef11bc31bcf212cf9/python/pyiceberg/cli/console.py#L92-L103

Seems it relies on `catalog.list_namespaces` to return an emptry list or not to determine whether `list_tables` should be called. I think this is the reason that we do not throw exception in `list_namespaces` in hive and glue.
Sorry that I did not realize this initially.

Fokko replied:
It doesn't seem to work at all:
```sh
➜  python git:(support-ddb-catalog) time pyiceberg --catalog dynamo list
∅
```

The hierarchical namespaces are just that you can have nested schemas: `finance.debitors.invoices.unpaid`. The CLI will first check if there is a nested namespace, and then if there are tables. For DynamoDB, you can just check if there are multiple levels (more than one dot), and if this is the case, just return an empty list:

```python
tuple_identifier = Catalog.identifier_to_tuple(identifier)
if len(tuple_identifier) > 1:
    return []
```

arminnajafi replied:
@JonasJ-ap 
Great! Thanks. It makes sense now.

arminnajafi replied:
@Fokko 
Not sure what happened for you there but I wasn't able to reproduce your issue. It seemed to work for me:
```
(pyiceberg-py3.10) ➜  python git:(support-ddb-catalog) ✗ time pyiceberg --catalog dynamo list
nyc
pyiceberg --catalog dynamo list  1.25s user 1.91s system 243% cpu 1.299 total
(pyiceberg-py3.10) ➜  python git:(support-ddb-catalog) ✗ 

```

arminnajafi replied:
But I updated the logic to `if len(tuple_identifier) > 1:` anyway which I think is more descriptive and accurate to what we had before.

Fokko replied:
The error appeared when `time pyiceberg --catalog dynamo list nyc` was run. It looks good now, thanks for fixing this 👍🏻

arminnajafi commented:
Thanks for the review @Fokko !

I addressed most of the comments in the new commits but also asked a few questions.
Let me know whenever you get a chance to go over them. 
It might be easier to discuss some of them in Slack.

Fokko reviewed (approved):
LGTM, thanks for working on this @arminnajafi

jackye1995 reviewed (approved):
looks good to me! Thanks for all the refactoring!

JonasJ-ap reviewed (commented):
Overall LGTM! Thanks for all your work. I have a followup regarding to the changes made to the `list_namespaces` method

JonasJ-ap commented on python/pyiceberg/catalog/glue.py:
```diff
@@ -461,8 +390,10 @@ def list_namespaces(self, namespace: Union[str, Identifier] = ()) -> List[Identi
         # Glue does not support hierarchical namespace, therefore return an empty list
-        if namespace:
+        tuple_identifier = Catalog.identifier_to_tuple(namespace)
+        if len(tuple_identifier) > 1:
```
For example:
If I run `pyiceberg list default`, now I will get all my databases:
```
(pyiceberg-py3.9) (base) ➜  python git:(support-ddb-catalog) pyiceberg list default
default                                   
delta_lake_demo                           
delta_playgrounddb                        
hudi_playgrounddb                         
migrate_delta_to_iceberg_demo             
migrated_to_iceberg_demo                  
mygluedatabase                            
mygluedatabase2                           
...
```
but previously I will get all my tables in `default` database:
```
(pyiceberg-py3.9) (base) ➜  python git:(master) pyiceberg list default
default.cloudtrail_logs_aws_cloudtrail_logs_481640105715_2dd430ba
default.cloudtrail_logs_aws_cloudtrail_logs_481640105715_de0525cc
default.table100                                                 
default.test2                                                    
default.test21                                                   
default.test22                                                   
default.test23                                                   
default.test24                                                   
default.test27                                                   
default.test28                                                   
default.test29                                                   
default.test3     
...
```

arminnajafi replied:
Good catch. Changed it to what it was before:
```
if namespace:
    return []
```

arminnajafi commented on python/tests/catalog/integration_test_dynamodb.py:
```diff
@@ -0,0 +1,258 @@
+    for database_name in database_list:
+        assert (database_name,) in db_list
+
+    with pytest.raises(ValidationError, match="This API is not supported for hierarchical namespaces."):
```
Good catch. Fixed. I completely missed checking integ tests and GitHub actions didn't run them either.

Fokko commented:
Merging this, thanks for the awesome work @arminnajafi! And thanks @jackye1995 @JonasJ-ap for the reviews, much appreciated!
